### PR TITLE
[master] Add bech32 utilities and initial migration to bech32 address

### PR DESCRIPTION
**Pull request by sunny-teo9000**
## Description
To add Bech32 utilities and allow the following API to accept both base16 and bech32 for now.

GetBalance
GetSmartContractCode
GetSmartContractInit
GetSmartContracts
GetSmartContractState
GetSmartContractSubState

## Backward Compatibility

- [x] This is not a breaking change
- [ ] This is a breaking change

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)
Was tested with a local isolated server, testing with the bech32 and base16 address

- [x] local machine test

- [ ] small-scale cloud test

### CODE REVIEW

File: src/libUtils/Bech32.cpp [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+}
+
+bool Bech32::isZilBase16(const std::string& input) {
+  return std::regex_match(input, BASE16_ZIL_REGEX_FORMAT);
+}
+
+bool Bech32::HasZilHrp(const std::string& input) {
+  if (input.length() < 4) return false;
+
+  auto hrp = boost::string_view(input);
```
Comment (KaustubhShamshery) [inline src/libUtils/Bech32.cpp:195]: const auto& hrp

File: src/libUtils/Bech32.cpp [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+  bool lower = false;
+  bool upper = false;
+
+  size_t inputSize = input.length();
+  if (inputSize < 8 || inputSize > 90) {
+    LOG_GENERAL(WARNING, "Cannot decode bech32 with invalid size: " << input);
+    return "";
+  }
+
+  for (size_t i = 0; i < inputSize; ++i) {
```
Comment (KaustubhShamshery) [inline src/libUtils/Bech32.cpp:210]: modernize loop

File: src/libUtils/Bech32.cpp [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+    char c = input[i];
```
Comment (KaustubhShamshery) [inline src/libUtils/Bech32.cpp:211]: can use const auto& here too

File: src/libUtils/Bech32.cpp [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+
+#include <algorithm>
+#include <boost/algorithm/string.hpp>
+#include <boost/utility/string_view.hpp>
+#include <iostream>
+#include <regex>
+
+#include "Bech32.h"
+#include "libServer/AddressChecksum.h"
+
```
Comment (KaustubhShamshery) [inline src/libUtils/Bech32.cpp:26]: can use using namespace std in .cpp file

File: src/libUtils/Bech32.cpp [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+const std::string Bech32::ZIL_HRP = "zil";
+
+const char Bech32::HEX_ARRAY[16] = {'0', '1', '2', '3', '4', '5', '6', '7',
+                                    '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
+
+const std::regex Bech32::BECH32_ZIL_REGEX_FORMAT(
+    "^zil1[qpzry9x8gf2tvdw0s3jn54khce6mua7l]{38}");
+const std::regex Bech32::BASE16_ZIL_REGEX_FORMAT("^(0x)?[0-9a-fA-F]{40}");
+
+int Bech32::hexCharToInt(char hexChar) {
```
Comment (ansnunez) [inline src/libUtils/Bech32.cpp:47]: Using streams to convert individual characters might be inefficient. Something like this should suffice:
```
int i = -1;
if (hexChar >= '0' && hexChar <= '9') {
  i = hexChar - '0';
} else if (hexChar >= 'a' && hexChar <= 'f') {
  i = 10 + hexChar - 'a';
}
return i;
```

Comment (ansnunez) [reply src/libUtils/Bech32.cpp:47]: Additionally, if you also check for `(hexChar >= 'A' && hexChar <= 'F')` then `hexStringToCharVec` won't need to call `boost::to_lower_copy` anymore

File: tests/Utils/Test_Bech32.cpp [hunk 6]
```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (C) 2019 Zilliqa
```
Comment (ansnunez) [inline tests/Utils/Test_Bech32.cpp:2]:
```suggestion
 * Copyright (C) 2021 Zilliqa
```

File: tests/Utils/Test_Bech32.cpp [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+  for (auto& pair : testPairs) {
+    bool result = boost::iequals(
+        Bech32::EncodeZilFromBase16ToBase32(pair.first), pair.second);
+    BOOST_CHECK_MESSAGE(result, "Wrong encode result for " + pair.first +
+                                    ", expecting: " + pair.second);
+  }
+}
+
+BOOST_AUTO_TEST_SUITE_END()
```
Comment (ansnunez) [inline tests/Utils/Test_Bech32.cpp:143]: Can you also add some tests for negative/failure cases?

**Review by ansnunez — CHANGES_REQUESTED** [review]

**Review by ansnunez — APPROVED** [review]

**Comment by bb111189** [comment]
https://github.com/Zilliqa/cryptoutils/tree/main/lib

Can we use bech32 from cryptoutils? This is to avoid 2 copy of bech32. @vaivaswatha  has recently created it and will be using for his VM. I think on the core side, ethash is already using cryptoutils

**Comment by vaivaswatha** [comment]
> https://github.com/Zilliqa/cryptoutils/tree/main/lib
> 
> Can we use bech32 from cryptoutils? This is to avoid 2 copy of bech32. @vaivaswatha has recently created it and will be using for his VM. I think on the core side, ethash is already using cryptoutils

Example usage: https://github.com/Zilliqa/scilla-vm/blob/4c58dae89aa3bc1567f21ad1ad9aa952ee089142/libsrtl/ScillaBuiltins.cpp#L608. This function and the next one do decoding and encoding respectively.

File: src/libServer/LookupServer.cpp [hunk 8]
```diff
@@ -40,9 +43,59 @@ using namespace std;
 CircularArray<std::string> LookupServer::m_RecentTransactions;
 std::mutex LookupServer::m_mutexRecentTxns;
 
+namespace {
 const unsigned int PAGE_SIZE = 10;
 const unsigned int NUM_PAGES_CACHE = 2;
 const unsigned int TXN_PAGE_SIZE = 100;
+const unsigned UINT8_ADDR_SIZE = ACC_ADDR_SIZE * 2;
+
+bool HasZilHrp(const string& input) {
```
Comment (ansnunez) [inline src/libServer/LookupServer.cpp:52]: I think this could actually be reduced to just
```
inline bool HasZilHrp(const string& input) {
   return input.substr(0, 4) == string("zil1");
}
```
If above is used, then the boost hpps can also be un-included

File: src/libServer/LookupServer.cpp [hunk 9]
```diff
@@ -40,9 +43,59 @@ using namespace std;
 CircularArray<std::string> LookupServer::m_RecentTransactions;
 std::mutex LookupServer::m_mutexRecentTxns;
 
+namespace {
 const unsigned int PAGE_SIZE = 10;
 const unsigned int NUM_PAGES_CACHE = 2;
 const unsigned int TXN_PAGE_SIZE = 100;
+const unsigned UINT8_ADDR_SIZE = ACC_ADDR_SIZE * 2;
```
Comment (ansnunez) [inline src/libServer/LookupServer.cpp:50]:
```suggestion
const unsigned int UINT8_ADDR_SIZE = ACC_ADDR_SIZE * 2;
```

**Review by ansnunez — CHANGES_REQUESTED** [review]

File: src/libServer/LookupServer.cpp (hunk 8 above)
Comment (sunny-teo9000) [reply src/libServer/LookupServer.cpp:52]: the boost string_view is actually a read-only wrapper to allow substring without constructing a new string.
But I guess since this is a short string, it should be ok. Will reduce this. :)

File: src/libServer/LookupServer.cpp [hunk 10]
```diff
@@ -40,9 +41,55 @@ using namespace std;
 CircularArray<std::string> LookupServer::m_RecentTransactions;
 std::mutex LookupServer::m_mutexRecentTxns;
 
+namespace {
 const unsigned int PAGE_SIZE = 10;
 const unsigned int NUM_PAGES_CACHE = 2;
 const unsigned int TXN_PAGE_SIZE = 100;
+const unsigned int UINT8_ADDR_SIZE = ACC_ADDR_SIZE * 2;
```
Comment (KaustubhShamshery) [inline src/libServer/LookupServer.cpp:48]: should be HEX_ADDR_SIZE

**Comment by KaustubhShamshery** [comment]
Can we have the unit test again?

**Review by ansnunez — APPROVED** [review]

**Review by KaustubhShamshery — APPROVED** [review]
